### PR TITLE
programs.nix-your-shell: add module

### PR DESCRIPTION
**Pull request by terlar**
### Description

This is alternative to #4645 which stalled. More along the lines of how starship and zoxide is implemented.

Closes #4645

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://nix-community.github.io/home-manager/#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/programs/nix-your-shell.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  cfg = config.programs.nix-your-shell;
+
+in {
+  meta.maintainers = [ maintainers.terlar ];
+
+  options.programs.nix-your-shell = {
+    enable = mkEnableOption ''
+      {command}`nix-your-shell`, a wrapper for `nix develop` or `nix-shell`
+      to retain the same shell inside the new environment.
+    '';
```
Comment (sumnerevans) [inline modules/programs/nix-your-shell.nix:16]: the trailing period is added by `mkEnableOption`

```suggestion
      to retain the same shell inside the new environment'';
```

File: tests/modules/programs/nix-your-shell/enable-shells.nix [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      enableFishIntegration = true;
+      enableNushellIntegration = true;
+      enableZshIntegration = true;
+    };
+    fish.enable = true;
+    nushell.enable = true;
+    zsh.enable = true;
+  };
+
+  # Needed to avoid error with dummy fish package.
```
Comment (sumnerevans) [inline tests/modules/programs/nix-your-shell/enable-shells.nix:16]: where is `fish` being stubbed?

**Review by sumnerevans — CHANGES_REQUESTED** [review]

File: tests/modules/programs/nix-your-shell/enable-shells.nix (hunk 2 above)
Comment (terlar) [reply tests/modules/programs/nix-your-shell/enable-shells.nix]: Seems fish ran into other issues when being stubbed (complain about `fish_indent`), and I didn't see any other tests stubbing it, so I removed that.

File: modules/programs/nix-your-shell.nix (hunk 1 above)
Comment (sumnerevans) [reply modules/programs/nix-your-shell.nix]: putting `''` on the same line was intentional (a space is added before the `.` otherwise)

Comment (terlar) [reply modules/programs/nix-your-shell.nix]: Sorry, missed that, pushed a fix.

**Comment by terlar** [comment]
@sumnerevans sorry to waste your time with the previous roundtrip and unfixed suggestion. Hopefully all should be good now.

**Review by dawidd6 — APPROVED** [review]

**Comment by terlar** [comment]
@rycee Do you think we could get this merged?

**Comment by rycee** [comment]
Thanks! Finally merged to master now 🙂

**Pull request merged**